### ISSUE TITLE
`Result::and` should check if the inner type has an `and` method

### ISSUE THREAD

**Issue by connortsui20** [rust-lang/rust-clippy#16847]
### What it does

This is the first time I'm making an issue for clippy, so apologies if this is either the wrong place or it isn't really suited to be a lint.

I recently ran into something similar to this toy example:

[_Playground Link_](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=3784c7a68153b8227172621037ed9f2e)

```rust
#[derive(Debug, Clone)]
struct AndType {
    inner: u8,
}

impl AndType {
    fn new(x: u8) -> Self {
        Self { inner: x }
    }

    fn and(self, other: Self) -> Self {
        Self::new(self.inner & other.inner)
    }
}

fn foo(x: u8) -> Result<AndType, ()> {
    Ok(AndType::new(x))
}

fn correct() -> Result<AndType, ()> {
    Ok(foo(0b0001)?.and(foo(0b1111)?))
}

fn foot_gun() -> Result<AndType, ()> {
    // Forgot to add the `?`s here. What could possibly go wrong?
    foo(0b0001).and(foo(0b1111))
}

fn main() {
    // We should expect `0b0001` as the result.
    let correct_res = correct().unwrap().inner;
    println!("{:04b}", correct_res);

    // Looks similar enough...
    let footgun_res = foot_gun().unwrap().inner;
    println!("{:04b}", footgun_res);
    assert_eq!(correct_res, footgun_res); // Panics as `footgun_res == 0b1111`
}
```

The issue here is pretty obvious in this example, where we forget to `?` the `Result` of `foo` and instead of calling `AndType::and`, we call [`Result::and`](https://doc.rust-lang.org/std/result/enum.Result.html#method.and).[^1] However, hopefully you can imagine that a more complex situation can exist where this is much harder to find.

The exact situation I ended up in was where I had 2 `Result<BitVector>`s that I was `and`ing together. I forgot to `?` the function calls that returned the `BitVector`s, so my first `BitVector` was completely thrown away.[^2]

Again, this seems easy to find in retrospect, but that function that returned the `Result<BitVector>` was only recently refactored from returning just a `BitVector`, so this was a silent change that was not caught by the compiler (the outer calling function was changed to `Result` without much thought).

So I think that a new lint would be nice here so that nobody else runs into a similar issue. It should check when `Result::and` is called with method syntax, and if the inner type has an `and` method, the lint should propose using qualified syntax instead.

```rust
fn arguably_less_wrong() -> Result<AndType, ()> {
    Result::and(foo(0b0001), foo(0b1111))
}
```

[^1]: To be honest, I have no idea why the `and` method is called "and" in the first place (who needed that first `Result` anyways...)
[^2]: Note that I wasn't actually working with a type called `BitVector`, it was actually a validity mask similar to a null map on a SQL column, but `BitVector` is easier to understand here.

### Advantage

This can prevent footguns where the programmer might think they are calling the inner `and` method, when in reality they are calling `Result::and`.

### Drawbacks

Depending on where this lint ends up (I am not super familiar with the different Rust lint levels), this could be a breaking change?

But at the same time, my bias here is that `Result::and` has incredibly unexpected behavior (as opposed to `and_then`, which not only actually uses the first result value if it is `Ok`, but also takes a function closure that needs to be typed correctly).

### Example

```rust
fn foo(x: u8) -> Result<AndType, ()> {
    Ok(AndType::new(x))
}

fn foot_gun() -> Result<AndType, ()> {
    foo(0b0001).and(foo(0b1111))
}
```

Could be written as:

```rust
fn foot_gun() -> Result<AndType, ()> {
    Result::and(foo(0b0001), foo(0b1111))
}
```

### Comparison with existing lints

I don't think that there are any lints for `Result::and` right now.

### Additional Context

I am happy to make a contribution with some guidance (I have never contributed to clippy before).

Note that I think the reason this lint makes sense on `Result::and` and not something like `Option::and` is just that the behavior of `Result::and` is so unexpected to me. An alternative solution to this problem would be to rename `Result::and` to something more fitting (like `Result::ok_and` or something).

I realize that this is supposed to be the opposite of `Result::or`, but for some reason `Result::and` just feels a lot different (or in other words, the short circuiting behavior doesn't make a lot of sense - something something bifunctor / bimonad).

**Comment by Souradip121** [comment]
@rustbot claim

I've implemented a result_and_inner_method lint in the suspicious category that addresses this footgun. The lint fires when .and() is called on a Result<T, E> where the inner type T has an inherent method also named and. It uses get_adt_inherent_method from clippy_utils to inspect the inner type at compile time, so there are no false positives for common cases like Result<u8, E> where u8 has no and method.

The suggestion rewrites foo.and(bar) to Result::and(foo, bar), making the call site explicit about which and is being invoked. The lint is warn-by-default and only checks inherent methods on the inner type, not trait methods, which keeps the scope conservative and avoids surprising users.
